hazelcast/src/main/java/com/hazelcast/sql/impl/type/converter/ShortConverter.java

```diff
@@ -38,7 +40,14 @@ private ShortConverter() {
 
     @Override
     public byte asTinyint(Object val) {
-        return (byte) cast(val);
+        short casted = cast(val);
+        byte converted = (byte) casted;
+
+        if (converted != casted) {
+            throw cannotConvert(QueryDataTypeFamily.TINYINT, val);
```

Comment:
This code produces a generic data exception. We will not be able to map it properly to SQLSTATE codes for JDBC in the future. ANSI SQL declares a family of error codes for data exceptions, with `numeric value out of range` being a special separate code. 
I am not sure that we should introduce many sub-codes for `data exception` at the moment, because we do not have JDBC. But at the very least the error message should reflect that overflow happened. This is not the case now since we throw an abstract `Cannot convert ...`, without explaining why the conversion is not possible.

Reply:
fixed

Reply:
In other databases, it is often referred to as "numeric overflow", to emphasize that we are talking about numbers. SQLSTATE also has the word "numeric" AFAIU. 
So I would propose to add the word "numeric" to the exception message as well.

Reply:
fixed